<h1>SQL <span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Анализ-данных-сервиса-для-чтения-книг-по-подписке" data-toc-modified-id="Анализ-данных-сервиса-для-чтения-книг-по-подписке-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Анализ данных сервиса для чтения книг по подписке</a></span><ul class="toc-item"><li><span><a href="#Описание-кейса" data-toc-modified-id="Описание-кейса-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание кейса</a></span></li><li><span><a href="#Цель-и-задачи-исследования" data-toc-modified-id="Цель-и-задачи-исследования-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Цель и задачи исследования</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Разведочный-анализ-данных" data-toc-modified-id="Разведочный-анализ-данных-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Разведочный анализ данных</a></span></li></ul></li></ul></div>

# Анализ данных сервиса для чтения книг по подписке

## Описание кейса

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Цель и задачи исследования

**Цель**: проанализировать базу данных крупного сервиса для чтения книг по подписке для формулирования ценностное предложения для нового продукта.

**Задачи**:
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных


**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


ER-диаграмма базы данных

![Картинка](Untitled_(33).png)

## Разведочный анализ данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Посмотрим на первые 5 строк каждой из таблиц:

In [3]:
table_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [4]:
def five_rows(table):
    query = """
    SELECT *
    FROM {}
    LIMIT 5
    """.format(table)
    df = pd.io.sql.read_sql(query, con = engine) 
    display(df)

In [5]:
for table in table_list:
    print('Таблица - ' + table)
    five_rows(table)
    print('---------------------------------------------')

Таблица - books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

---------------------------------------------
Таблица - authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

---------------------------------------------
Таблица - publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

---------------------------------------------
Таблица - ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

---------------------------------------------
Таблица - reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

---------------------------------------------


Теперь выполним задания:

1. Посчитайте, сколько книг вышло после 1 января 2000 года;

In [6]:
query_1 = """
SELECT 
    COUNT(*)
FROM books
WHERE CAST(publication_date AS date) > '2000-01-01'::date
"""

In [7]:
df_1 = pd.io.sql.read_sql(query_1, con = engine) 

In [8]:
df_1

count
0    819

Всего таких книг вышло 819 штук

2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [21]:
query_2 = """
WITH score AS (
    SELECT 
        r.book_id,
        AVG(r.rating::numeric)as mean_score
    FROM ratings r
    GROUP BY 1),
reviews_data AS (
    SELECT 
        book_id,
        COUNT(*) as count_reviews
    FROM reviews
    GROUP BY 1)
SELECT 
     s.book_id,
     b.title,
     s.mean_score,
     rd.count_reviews
FROM score s
LEFT JOIN reviews_data rd ON rd.book_id = s.book_id
LEFT JOIN (SELECT 
                book_id,
                title
           FROM books) b ON b.book_id = s.book_id
ORDER BY s.mean_score DESC;
"""

In [22]:
df_2 = pd.io.sql.read_sql(query_2, con = engine) 

In [23]:
df_2

book_id                                              title  mean_score  \
0         86      Arrows of the Queen (Heralds of Valdemar  #1)        5.00   
1        901  The Walking Dead  Book One (The Walking Dead #...        5.00   
2        390                                    Light in August        5.00   
3        972  Wherever You Go  There You Are: Mindfulness Me...        5.00   
4        136  Captivating: Unveiling the Mystery of a Woman'...        5.00   
..       ...                                                ...         ...   
995      915  The World Is Flat: A Brief History of the Twen...        2.25   
996      316                  His Excellency: George Washington        2.00   
997      202                                      Drowning Ruth        2.00   
998      371                                              Junky        2.00   
999      303                               Harvesting the Heart        1.50   

     count_reviews  
0              2.0  
1              2.0  
2              2.0  
3              2.0  
4              2.0  
..             ...  
995            3.0  
996            2.0  
997            3.0  
998            2.0  
999            2.0  

[1000 rows x 4 columns]

Всего в выборке 1000 книг. Максимальная оценка, логично, - 5, минимальная - 1

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [12]:
query_3 = """
WITH cte as (SELECT 
                publisher_id,
                COUNT(*) as count_books
            FROM books
            WHERE num_pages > 50
            GROUP BY 1)
SELECT 
    c.publisher_id,
    p.publisher
FROM cte c
LEFT JOIN publishers p ON p.publisher_id = c.publisher_id
WHERE count_books = (SELECT MAX(count_books)
                    FROM cte)

"""

In [13]:
df_3 = pd.io.sql.read_sql(query_3, con = engine) 

In [14]:
df_3

publisher_id      publisher
0           212  Penguin Books

Таким издателельством является Penguin Books

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [24]:
query_4 = """
WITH books_stat AS (
    SELECT 
        r.book_id,
        COUNT(*) as count_rate,
        AVG(r.rating::numeric) as mean_score
    FROM ratings r
    GROUP BY 1
    HAVING COUNT(*) > 50),
author_data AS (
    SELECT bs.book_id,
           bs.mean_score,
           b.author_id
    FROM books_stat bs
    LEFT JOIN (SELECT 
                    book_id,
                    author_id
               FROM books) b ON b.book_id = bs.book_id),
author_score AS (    
    SELECT author_id,
           AVG(mean_score) AS mean_score_author
    FROM author_data
    GROUP BY 1
    ORDER BY 2 DESC)
    
SELECT us.author_id,
       a.author
FROM author_score us
LEFT JOIN authors a ON a.author_id = us.author_id
WHERE us.mean_score_author = (SELECT MAX(mean_score_author)
                              FROM author_score)
"""

In [25]:
df_4 = pd.io.sql.read_sql(query_4, con = engine) 

In [26]:
df_4

author_id                      author
0        236  J.K. Rowling/Mary GrandPré

Здесь конечно же всеми любимая Джоан Роалинг.

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [27]:
query_5 = """
WITH users_stat AS (
    SELECT 
        username,
        COUNT(*) AS count_rate
    FROM ratings
    GROUP BY 1
    HAVING COUNT(*) > 50)

SELECT round(AVG(d.count_reviws)) as mean_reviews
FROM (SELECT 
        r.username,
        COUNT(*) AS count_reviws   
    FROM reviews r
    WHERE EXISTS (SELECT 1
                 FROM users_stat us 
                 WHERE us.username = r.username)
    GROUP BY 1) as d
"""

In [28]:
df_5 = pd.io.sql.read_sql(query_5, con = engine) 

In [29]:
df_5

mean_reviews
0          24.0

Среднее количество обзоров у данныхпользователей - 24